### ISSUE TITLE
Thrown errors are cached and re-thrown rather than retrying

### ISSUE THREAD

**Issue by onionhammer** [tannerlinsley/react-query#662]
**Describe the bug**
Relating to this comment, when retry failed due to error it appears that the error is being cached forever and rethrown every time the query itself is retried (not the internal retry logic)

https://github.com/tannerlinsley/react-query/issues/638#issuecomment-651210297

**To Reproduce**
1. Create a global `retry` handler that returns false if there is an `err` argument
2. After the error is thrown, handle it with an error boundary so the next time the query is run the query should succeed

**Expected behavior**
The query should re-run and no error should occur

**What actually happens**
the error is cached and rethrown

**Comment by marcosvega91** [comment]
Hi @onionhammer thanks for opening the issue :).

Maybe I don't understand you question well, but If you return `false` from your custom function the query will not re-run. You should return `true` to re-run the query or you could check `err` argument and return `true|false` only for particular cases.

**Comment by onionhammer** [comment]
@marcosvega91 the issue is if I add in custom logic to handle a 401 error (the user isnt authorized), and redirect them to the login screen, the component that is querying the resource which 401's will be unloaded, then when they login and come back, the query has a cached error. This is a breaking change/error in 2.x range, not present in 1.5x range of react-query

**Comment by marcosvega91** [comment]
ok,  I understand. Could you please create a case on [codesandbox](https://codesandbox.io/), in this way I can take a look at your code? thanks

**Comment by onionhammer** [comment]
@marcosvega91  here's some code that reproduces the issue:

https://codesandbox.io/s/bitter-bash-13bvc?file=/src/App.js

Note if you start at the /login route and click login, it works, but if the query tries to load the data initially it will always have that error

**Comment by marcosvega91** [comment]
I have looked at your code and I think that the problem could be the `suspense` because the cache will be cleaned on component unmount by `scheduleGarbageCollection`. BTW with `suspense` `useEffect` is not called as his callback too
```javascript
// Subscribe to the query when the subscribe function changes
  React.useEffect(() => {
    instanceRef.current = query.subscribe(() => rerender({}))

    // Unsubscribe when things change
    return () => instanceRef.current.unsubscribe()
  }, [query, rerender])
```

**Comment by onionhammer** [comment]
@marcosvega91 it was cleared in previous versions of react-query, but not 2.x

**Comment by marcosvega91** [comment]
Your code seams right, I think it's a bug. @tannerlinsley What do you think ?

**Comment by tannerlinsley** [comment]
Yeah, it seems like a bug.

**Comment by tannerlinsley** [comment]
This will be fixed in #840
